ipn/ipnlocal/profiles.go

```
@@ -0,0 +1,467 @@
+	if !ok {
+		return ErrProfileNotFound
+	}
+
+	if pm.currentProfile != nil && kp.Key == pm.currentProfile.Key && pm.prefs.Valid() {
+		return nil
+	}
+	if kp.LocalUserID != pm.currentUserID {
+		return fmt.Errorf("profile %q is not owned by current user", profile)
+	}
+	prefs, err := pm.loadSavedPrefs(kp.Key)
+	if err != nil {
+		return err
+	}
+	pm.prefs = prefs
+	pm.currentProfile = kp
+	pm.isNewProfile = false
+	return pm.setAsUserSelectedProfileLocked()
+}
+
+func (pm *ProfileManager) setAsUserSelectedProfileLocked() error {
+	k := ipn.CurrentProfileKey(pm.currentUserID)
+	if pm.currentProfile == nil {
+		return pm.store.WriteState(k, nil)
+	}
+	return pm.store.WriteState(k, []byte(pm.currentProfile.Key))
+}
+
+func (pm *ProfileManager) loadSavedPrefs(key ipn.StateKey) (ipn.PrefsView, error) {
+	bs, err := pm.store.ReadState(key)
+	if err != nil {
+		if err == ipn.ErrStateNotExist {
+			return emptyPrefs, nil
+		}
+		return ipn.PrefsView{}, err
+	}
+	savedPrefs, err := ipn.PrefsFromBytes(bs)
+	if err != nil {
+		return ipn.PrefsView{}, fmt.Errorf("PrefsFromBytes: %v", err)
+	}
+	pm.logf("using backend prefs for %q: %v", key, savedPrefs.Pretty())
+
+	// Ignore any old stored preferences for https://login.tailscale.com
+	// as the control server that would override the new default of
+	// controlplane.tailscale.com.
+	if savedPrefs.ControlURL != "" &&
+		savedPrefs.ControlURL != ipn.DefaultControlURL &&
+		ipn.IsLoginServerSynonym(savedPrefs.ControlURL) {
+		savedPrefs.ControlURL = ""
+	}
+	return savedPrefs.View(), nil
+}
+
+// CurrentProfile returns the name and ID of the current profile, or "" if the profile
+// is not named.
+func (pm *ProfileManager) CurrentProfile() ipn.LoginProfile {
+	pm.mu.Lock()
+	defer pm.mu.Unlock()
+	if pm.currentProfile == nil {
+		return ipn.LoginProfile{}
+	}
+	return *pm.currentProfile
+}
+
+var ErrProfileNotFound = errors.New("profile not found")
+
+// DeleteProfile removes the profile with the given id. It is a no-op if the
```

It does return an error though, so a caller may end up stopping instead of continuing if the profile does not exist.

---

Updated docs.